x/gov/types/msgs.go

```diff
@@ -46,6 +62,19 @@ func (m *MsgSubmitProposal) GetProposer() sdk.AccAddress {
 	return proposer
 }
 
+func (m *MsgSubmitProposal) GetMessages() ([]sdk.Msg, error) {
+	msgs := make([]sdk.Msg, len(m.Messages))
+	for i, msgAny := range m.Messages {
+		msg, ok := msgAny.GetCachedValue().(sdk.Msg)
+		if !ok {
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "messages contains %T which is not a sdk.Msg", msgAny)
```

Comment:
Getting this error message when I run `TestSimulateMsgSubmitProposal`:
```
messages contains *types.Any which is not a sdk.Msg: invalid request
```
I've just copied the same format that was applied to `Content` with `sdk.Msg`. Is there something I might be missing?

Reply:
seems like you also need to update `MsgSubmitProposal.UnpackInterfaces` to unpack the msgs (https://github.com/cosmos/cosmos-sdk/pull/9810#discussion_r678898995)

Reply:
Ah thanks, I missed that.

Reply:
Am I able to unpack both the content and the messages interface together?

Reply:
I'm currently getting the problem where this
https://github.com/cosmos/cosmos-sdk/blob/9c4193c29ece26970651e257d869ec5a2063675c/x/gov/simulation/operations_test.go#L122
is failing. When I log the msg after creating it here https://github.com/cosmos/cosmos-sdk/blob/9c4193c29ece26970651e257d869ec5a2063675c/x/gov/simulation/operations.go#L142
It seems to be okay but in between packing and unpacking, the data gets lost and nil is returned

Reply:
> Am I able to unpack both the content and the messages interface together?

Not really, it seems like content is not getting unpacked because unpacking the msgs returns an error before it can get to it: `Unregistered interface types.Msg`. This is coming from amino. You might need to register this interface in `ModuleCdc`.

Reply:
Are you sure it's not already registered?
I get `panic: TypeInfo already exists for types.Msg` when adding it to:
```
func RegisterLegacyAminoCodec(cdc *codec.LegacyAmino) {
	cdc.RegisterInterface((*Content)(nil), nil)
	cdc.RegisterInterface((*sdk.Msg)(nil), nil)
	cdc.RegisterConcrete(&MsgSubmitProposal{}, "cosmos-sdk/MsgSubmitProposal", nil)
	cdc.RegisterConcrete(&MsgDeposit{}, "cosmos-sdk/MsgDeposit", nil)
	cdc.RegisterConcrete(&MsgVote{}, "cosmos-sdk/MsgVote", nil)
	cdc.RegisterConcrete(&MsgVoteWeighted{}, "cosmos-sdk/MsgVoteWeighted", nil)
	cdc.RegisterConcrete(&TextProposal{}, "cosmos-sdk/TextProposal", nil)
}
```

Reply:
This seems to be because `RegisterLegacyAminoCodec` in run from `simapp` `MakeTestEncodingConfig()` (through `ModuleBasics.RegisterLegacyAminoCodec(encodingConfig.Amino)`) as part of `simapp.Setup` in `operations_test.go` `createTestApp` and the thing is that `sdk.Msg` interface is already registered in `MakeTestEncodingConfig()` using `std.RegisterLegacyAminoCodec()` before.

The other issue is that the test is not using `simapp.LegacyAmino()` codec but `types.ModuleCdc` so one way to fix this would be to to add `cdc.RegisterInterface((*sdk.Msg)(nil), nil)` to gov `codec.go` `init()` instead.